Implemented Webrtc Loopback Test for webrtc private to private

asmit27rai opened the pull request:
Test:
```bash
pytest tests/core/transport/test_*webrtc*.py -v
```

Nkovaturient committed:
feat(webrtc):revamp-setup+implement-test-suites

Nkovaturient committed:
Merge branch 'main' of https://github.com/Nkovaturient/py-libp2p into feat/add-webrtc-transport

Nkovaturient committed:
fix(webrtc): resolve lint and type declaration issues

sukhman-sukh committed:
Add message.proto and fix a few mypy-ci errors

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

sukhman-sukh committed:
Fix SDP and ICE message transfer

sukhman-sukh committed:
Fix ICE candidate exchange

sukhman-sukh committed:
Fix lint error in CI

Nkovaturient committed:
fix pyrefly typecheck in CI

sukhman-sukh committed:
Fix webrtc-direct proto and certmanager

sukhman-sukh committed:
Add dialer and listener to webrtc-direct

sukhman-sukh committed:
Add stream read/write for webrtc

sukhman-sukh committed:
Fix lint error in CI

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

sukhman-sukh committed:
Merge branch 'feat/add-webrtc-transport' into webrtc-side-branch

sukhman-sukh committed:
Fix lint

sukhman-sukh committed:
Merge pull request #1 from sukhman-sukh/webrtc-side-branch

Webrtc side branch

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

asmit27rai committed:
Refactor WebRTC Private-to-Private Listener to Match js-libp2p Pattern (#2)

* reimplement private-to-private listener to match js-libp2p and fix signaling

* Address_Comments

* Add test

Commit:
fix lint issues in CI + remove mock test suites

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

asmit27rai committed:
Implemented Webrtc Loopback Test for webrtc private to private

Nkovaturient commented on libp2p/transport/webrtc/connection.py:
```diff
@@ -234,9 +234,10 @@ def on_close() -> None:
-                        self._close_trio_channels, trio_token=self._trio_token
-                    )
+                    def _spawn_close():
+                        trio.lowlevel.spawn_system_task(self._close_trio_channels())
```
`_close_trio_channels()` is an async function, but it's being called directly in `spawn_system_task()` - should this be wrapped differently?
we should pass the func, not call it?

```python
def _spawn_close():
    trio.lowlevel.spawn_system_task(self._close_trio_channels)
```

Nkovaturient commented on libp2p/transport/webrtc/async_bridge.py:
```diff
@@ -143,7 +143,17 @@ async def close_peer_connection(self, peer_connection: RTCPeerConnection) -> Non
+                result = close_attr()
+                if hasattr(result, "__await__"):
+                    try:
+                        await aio_as_trio(result)
```
This exception handling in `await aio_as_trio(result)` catches all exceptions but then tries `close_attr()` again - which might mask underlying errors,  right? 
we should log the specific exception before falling back.

Nkovaturient reviewed (commented):
thanks for your efforts and support here,
kindly address these

Nkovaturient commented:
hey @asmit27rai , suggested few changes, your test implementation currently  uses mocks (MockDataChannel, MockPeerConnection) and doesn't exercise real WebRTC components. those mock tests only verify:
- `WebRTCRawConnection` initialization (passing mocks)
- Stream ID allocation, Stream creation and management
- Connection wrapper logic (basic functionality)
- Data flow through connection wrapper (not real WebRTC)

This is NOT sufficient - since they only verify the connection wrapper works with mocks
also I believe, a proper loopback test for WebRTC private-to-private should:

1. **Create two libp2p hosts in the same process** (loopback)
2. **Set up WebRTC transport on both hosts**
3. **Establish a connection between them** (simulating p2p connection)
4. **Test data transmission** over the WebRTC connection
5. **Test stream multiplexing in webrtc pvt-to-pvt(see here, ref)** (https://github.com/libp2p/py-libp2p/pull/637#issuecomment-3383408573)
6. **Test connection cleanup** and resource management

also, since circuit-relay needs to be implemented properly, for now we can work with Mock Signaling Stream, i,e, using a simple in-memory stream

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

asmit27rai committed:
feat: add WebRTC pvt-to-pvt loopback tests + fixes

- Add 14 comprehensive loopback tests (10 unit + 4 integration)
- Fix spawn_system_task call in connection.py
- Add exception logging in async_bridge.py

All tests passing. Addresses comments..

asmit27rai commented:
@Nkovaturient I addressed the comment.
Please have a look.
Thank You...

Commit:
feat(webrtc): register base transport, enrich udp hole punching, and tighten webrtc tests

Nkovaturient committed:
feat(webrtc): Refactor webRTC signalling & security flow with relay discovery fixes

Nkovaturient committed:
Merge branch 'libp2p:main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

seetadev committed:
Merge branch 'main' into feat/add-webrtc-transport

Nkovaturient commented:
> This is NOT sufficient - since they only verify the connection wrapper works with mocks also I believe, a proper loopback test for WebRTC private-to-private should:
> 
> 1. **Create two libp2p hosts in the same process** (loopback)
> 2. **Set up WebRTC transport on both hosts**
> 3. **Establish a connection between them** (simulating p2p connection)
> 4. **Test data transmission** over the WebRTC connection
> 5. **Test stream multiplexing in webrtc pvt-to-pvt(see here, ref)** ([feat: Adding stream state to the Network Stream libp2p/py-libp2p#637 (comment)](https://github.com/libp2p/py-libp2p/pull/637#issuecomment-3383408573))
> 6. **Test connection cleanup** and resource management
> 
> also, since circuit-relay needs to be implemented properly, for now we can work with Mock Signaling Stream, i,e, using a simple in-memory stream

CCing @seetadev to examine this test file on above terms and further pointers.

Nkovaturient committed:
Merge branch 'libp2p:main' into feat/add-webrtc-transport

Nkovaturient committed:
fix(webrtc): handle ExceptionGroup,  improve err handling and add relay tests

yashksaini-coder committed:
Merge branch 'libp2p:main' into loop_test

yashksaini-coder committed:
Merge branch 'feat/add-webrtc-transport' into loop_test

seetadev commented:
> > This is NOT sufficient - since they only verify the connection wrapper works with mocks also I believe, a proper loopback test for WebRTC private-to-private should:
> > 
> > 1. **Create two libp2p hosts in the same process** (loopback)
> > 2. **Set up WebRTC transport on both hosts**
> > 3. **Establish a connection between them** (simulating p2p connection)
> > 4. **Test data transmission** over the WebRTC connection
> > 5. **Test stream multiplexing in webrtc pvt-to-pvt(see here, ref)** ([feat: Adding stream state to the Network Stream libp2p/py-libp2p#637 (comment)](https://github.com/libp2p/py-libp2p/pull/637#issuecomment-3383408573))
> > 6. **Test connection cleanup** and resource management
> > 
> > also, since circuit-relay needs to be implemented properly, for now we can work with Mock Signaling Stream, i,e, using a simple in-memory stream
> 
> CCing @seetadev to examine this test file on above terms and further pointers.


@Nkovaturient : Sincere apologies for the delay in response. We have been swamped up with resolving interop test issues.

Absolutely — this is a *great* breakdown, and we are completely aligned with you here. The current mock-only verification is a good starting point, but you’re right: it isn’t sufficient for true confidence in WebRTC private-to-private. A proper loopback test is exactly what will move us from “it compiles” to “it works end-to-end under real conditions.”

The checklist is spot-on, and honestly, it’s the kind of clarity that helps the whole effort move faster:

* **Two hosts in-process** gives us deterministic repeatability.

* **Full WebRTC transport setup on both** ensures we’re no longer testing abstractions but actual signaling + DTLS + SCTP behavior.

* **Connection establishment + real data transfer** is what validates the E2E handshake path.

* **Multiplexing tests** are important — especially with the new stream-state discussions in `#637`. This is where we’ll catch subtle channel/stream ordering issues before they reach users.

* **Cleanup + resource management** will tell us if we’re leaking channels, goroutines/tasks, or descriptors — super important for long-running agents.

* **Mock signaling stream for now** is perfect. With circuit-relay not fully wired yet, in-memory signaling keeps the test self-contained while still validating the whole WebRTC stack.

Great momentum — let’s push this across the finish line. Please re-run the CI/CD pipeline after approving the workflow.

yashksaini-coder commented:
@Nkovaturient Please approve the workflow so we can run CI checks

yashksaini-coder committed:
Merge branch 'libp2p:main' into loop_test

Nkovaturient commented:
ran the CI/CD checks, 
needs to be addressed.

1) adding `aiortc>=1.14.0`  & "trio-asyncio>=0.15.0" dependency in pyproject.toml [alternative: adding webrtc to dev extras so lint/typecheck environments (pre-commit) can resolve aiortc imports in test envs as well]
2) Fix pyrefly AsyncGenerator typing failures in tests
- The errors show EXPECTED Self@X <: AsyncGenerator[@, @] for functions that act as async generators/bridges. Add explicit return type annotations and ensure the typing import is present. 
- Apply this to the failing locations reported by pyrefly: `tests/core/transport/test_loopback_webrtc_pvt_to_pvt.py (lines ~141, 215, 266, 357)` and `tests/core/transport/test_webrtc_pvt_to_pvt_integration.py`. Adjust the precise signatures and element types to match the test implementation. 
> [Note: the lint run previously reported AsyncGenerator return-type expectations in a few test fixtures. If pyrefly still flags EXPECTED Self <: AsyncGenerator errors, add explicit typing to generator fixtures.]
3) The interop job fails while installing Nim. The step that runs: `curl -sSf https://nim-lang.org/choosenim/init.sh | sh -s -- -y --firstInstall returned HTTP 500 (curl exit code 22),` so the installer script download failed and the job aborted.

If anyone picks to resolve these issues, then inform beforehand to avoid duplicate efforts.

yashksaini-coder committed:
Merge branch 'libp2p:main' into loop_test

Nkovaturient commented:
> ran the CI/CD checks, needs to be addressed.
> 
> 1. adding `aiortc>=1.14.0`  & "trio-asyncio>=0.15.0" dependency in pyproject.toml [alternative: adding webrtc to dev extras so lint/typecheck environments (pre-commit) can resolve aiortc imports in test envs as well]
> 2. Fix pyrefly AsyncGenerator typing failures in tests
> 
> * The errors show EXPECTED Self@X <: AsyncGenerator[@, @] for functions that act as async generators/bridges. Add explicit return type annotations and ensure the typing import is present.
> * Apply this to the failing locations reported by pyrefly: `tests/core/transport/test_loopback_webrtc_pvt_to_pvt.py (lines ~141, 215, 266, 357)` and `tests/core/transport/test_webrtc_pvt_to_pvt_integration.py`. Adjust the precise signatures and element types to match the test implementation.
> 
> > [Note: the lint run previously reported AsyncGenerator return-type expectations in a few test fixtures. If pyrefly still flags EXPECTED Self <: AsyncGenerator errors, add explicit typing to generator fixtures.]
> 
> 3. The interop job fails while installing Nim. The step that runs: `curl -sSf https://nim-lang.org/choosenim/init.sh | sh -s -- -y --firstInstall returned HTTP 500 (curl exit code 22),` so the installer script download failed and the job aborted.
> 
> If anyone picks to resolve these issues, then inform beforehand to avoid duplicate efforts.

fixed the above CI/CD issues [here](https://github.com/libp2p/py-libp2p/pull/780/commits/f71e5245ed39eefc52cddc1131dbc486ff1f407d#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711)
